### PR TITLE
The check for internal types in RewriteArrayExprs

### LINKED ISSUE #5157: RewriteArrayExprs pass incorrectly rewrites all Array-returning operations
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Reproducer:
```python
from numba.extending import (
    overload,
    models,
    lower_builtin,
    register_model,
    make_attribute_wrapper,
    type_callable,
    typeof_impl,
)
from numba import types, cgutils
from numba import njit
import operator
import numpy as np


def make_dummy_type():
    class Dummy(object):
        def __init__(self, value):
            self.value = value

    class DummyType(types.Type):
        def __init__(self):
            super(DummyType, self).__init__(name="Dummy")

    dummy_type = DummyType()

    @register_model(DummyType)
    class DummyModel(models.StructModel):
        def __init__(self, dmm, fe_type):
            members = [("value", types.intp)]
            models.StructModel.__init__(self, dmm, fe_type, members)

    make_attribute_wrapper(DummyType, "value", "value")

    @type_callable(Dummy)
    def type_dummy(context):
        def typer(value):
            return dummy_type

        return typer

    @lower_builtin(Dummy, types.intp)
    def impl_dummy(context, builder, sig, args):
        typ = sig.return_type
        [value] = args
        dummy = cgutils.create_struct_proxy(typ)(context, builder)
        dummy.value = value
        return dummy._getvalue()

    @typeof_impl.register(Dummy)
    def typeof_dummy(val, c):
        return DummyType()

    # Store attributes
    return Dummy, DummyType


Dummy, DummyType = make_dummy_type()


@overload(operator.add)
def ov_add_dummy(lhs, rhs):
    if isinstance(lhs, DummyType) and isinstance(rhs, DummyType):
        print("===", lhs, rhs)

        def imp(lhs, rhs):
            return np.array([lhs.value, rhs.value])

        return imp


@njit
def foo(x, y):
    return Dummy(x) + Dummy(y)


r = foo(1, 2)
print(r)
```

The above raises:
```pytb
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython frontend)
No conversion from array(int64, 1d, C) to int64 for '$8return_value.3', defined at None

File "array_expr_issue.py", line 68:
def foo(x, y):
    return Dummy(x) + Dummy(y)
    ^

[1] During: typing of assignment at /Users/siu/dev/numba-git/array_expr_issue.py (68)

File "array_expr_issue.py", line 68:
def foo(x, y):
    return Dummy(x) + Dummy(y)
    ^

[1] During: lowering "$14binary_add.6 = arrayexpr(expr=(<built-in function add>, [Var($6call_function.2, array_expr_issue.py:68), Var($12call_function.5, array_expr_issue.py:68)]), ty=array(int64, 1d, C))" at /Users/siu/dev/numba-git/array_expr_issue.py (68)
```

Expected Behavior:
- `RewriteArrayExprs` pass should not make `arrayexpr` node for the `Dummy(x) + Dummy(y)` expression.
- Disabling the pass will allow the code run correctly.

### PR DESCRIPTION
**Pull request by Alexander-Makaryev**
Fixes #5157 

Please look at issue for more details.

The implementation is based on comment from @stuartarchibald :
https://github.com/numba/numba/issues/5157#issuecomment-699958032

### CODE REVIEW

File: numba/tests/test_array_exprs.py [hunk 1]
```diff
@@ -627,5 +644,127 @@ def do_call(x, y):
         self.assertTrue(isinstance(oty.type.dtype, types.Float))
 
 
+class TestExternalTypes(MemoryLeakMixin, unittest.TestCase):
+    """ Tests RewriteArrayExprs with external (user defined) types,
+    see #5157"""
+
+    source_lines = """
```
Comment (esc) [inline numba/tests/test_array_exprs.py:651]: would suggest to use `textwrap.dedent` for better code alignment here: https://docs.python.org/3/library/textwrap.html#textwrap.dedent

Comment (Alexander-Makaryev) [reply numba/tests/test_array_exprs.py]: done
thanks

**Comment by Alexander-Makaryev** [comment]
@stuartarchibald 
Looks like that we got some failure in CI tests that are not related to this PR.
Previous run of CI was green (https://dev.azure.com/numba/numba/_build/results?buildId=8051&view=results) and there are no significant changes in commits since that time.
What should I do? Do I need to report it somehow or restart tests in this PR?

**Comment by stuartarchibald** [comment]
@Alexander-Makaryev I've restarted the build, and am working on a fix, it's not this PR at fault. Once a fix for it makes it into mainline the CI builders will pick it up and it shouldn't happen again (at least for this particular issue).

File: numba/np/ufunc/array_exprs.py [hunk 2]
```diff
@@ -77,7 +77,14 @@ def _match_array_expr(self, instr, expr, target_name):
...
                 expr.fn in npydecl.supported_array_operators)):
             # It is an array operator that maps to a ufunc.
-            array_assigns[target_name] = instr
+            # check that all args have internal types
+            all_internal = True
+            for var in expr.list_vars():
+                if not self.typemap[var.name].is_internal:
+                    all_internal = False
+                    break
+            if all_internal:
```
Comment (stuartarchibald) [inline numba/np/ufunc/array_exprs.py:86]:
```suggestion
            # check that all args have internal types
            if all(self.typemap[var.name].is_internal
                   for var in expr.list_vars()):
```
think this is equivalent?

File: numba/tests/test_array_exprs.py [hunk 3]
```diff
@@ -627,5 +645,127 @@ def do_call(x, y):
...
+                f.write(lines)
+            # Add test_module to sys.path so it can be imported
+            sys.path.insert(0, tempdir)
+            test_module = importlib.import_module(temp_module_name)
+            yield test_module
+        finally:
+            sys.modules.pop(temp_module_name, None)
+            sys.path.remove(tempdir)
+            shutil.rmtree(tempdir)
+
```
Comment (stuartarchibald) [inline numba/tests/test_array_exprs.py:684]: It'd be nice to refactor this into `numba.tests.support`, but perhaps not for this PR.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch and for fixing this issue, just one minor comment to resolve else looks good.

File: numba/np/ufunc/array_exprs.py (hunk 2 above)
Comment (Alexander-Makaryev) [reply numba/np/ufunc/array_exprs.py]: @stuartarchibald 
done
thanks

File: numba/tests/test_array_exprs.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/tests/test_array_exprs.py:684]: https://github.com/numba/numba/issues/6778 tracks

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fix @Alexander-Makaryev, looks good.

**Pull request merged**